cirq-google/cirq_google/devices/named_topologies.py

```diff
@@ -0,0 +1,331 @@
+# Copyright 2021 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import abc
+import dataclasses
+import warnings
+from dataclasses import dataclass
+from typing import Dict, List, Tuple, Any, Sequence, Union, Iterable
+
+import cirq
+import networkx as nx
+from matplotlib import pyplot as plt
+
+
+def dataclass_json_dict(obj: Any, namespace: str = None) -> Dict[str, Any]:
+    return cirq.obj_to_dict_helper(
+        obj, [f.name for f in dataclasses.fields(obj)], namespace=namespace
+    )
+
+
+class NamedTopology(metaclass=abc.ABCMeta):
+    """A topology (graph) with a name.
+
+    "Named topologies" provide a mapping from a simple dataclass to a unique graph for categories
+    of relevant topologies. Relevant topologies may be hardware dependant, but common topologies
+    are linear (1D) and rectangular grid topologies.
+    """
+
+    name: str = NotImplemented
+    """A name that uniquely identifies this topology."""
+
+    n_nodes: int = NotImplemented
+    """The number of nodes in the topology."""
+
+    graph: nx.Graph = NotImplemented
+    """A networkx graph representation of the topology."""
+
+
+_GRIDLIKE_NODE = Union[cirq.GridQubit, Tuple[int, int]]
+
+
+def _node_and_coordinates(
+    nodes: Iterable[_GRIDLIKE_NODE],
+) -> Iterable[Tuple[_GRIDLIKE_NODE, Tuple[int, int]]]:
+    """Yield tuples whose first element is the input node and the second is guaranteed to be a tuple
+    of two integers. The input node can be a tuple of ints or a GridQubit."""
+    for node in nodes:
+        if isinstance(node, cirq.GridQubit):
+            yield node, (node.row, node.col)
+        else:
+            x, y = node
+            yield node, (x, y)
+
+
+def draw_gridlike(
+    graph: nx.Graph, ax: plt.Axes = None, cartesian: bool = True, **kwargs
+) -> Dict[Any, Tuple[int, int]]:
+    """Draw a Grid-like graph.
+
+    This wraps nx.draw_networkx to produce a matplotlib drawing of the graph.
+
+    Args:
+        graph: A NetworkX graph whose nodes are (row, column) coordinates.
+        ax: Optional matplotlib axis to use for drawing.
+        cartesian: If True, directly position as (row, column); otherwise,
```

Comment:
"Cartesian" isn't the best name for this parameter since a rotated Cartesian coordinate system is still a Cartesian coordinate system. Suggestion: "tilted". This better explains the meaning and is consistent with another name used elsewhere in the PR (`TiltedSquareLattice`).

Reply:
done